go/vt/vtctl/reparentutil/ers_sorter.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2021 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reparentutil
+
+import (
+	"sort"
+
+	"vitess.io/vitess/go/mysql"
+	"vitess.io/vitess/go/vt/vterrors"
+
+	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
+	vtrpcpb "vitess.io/vitess/go/vt/proto/vtrpc"
+)
+
+// ERSSorter sorts tablets by GTID positions and Promotion rules aimed at finding the best
+// candidate for intermediate promotion in emergency reparent shard
+type ERSSorter struct {
+	tablets   []*topodatapb.Tablet
+	positions []mysql.Position
+	idealCell string
+}
+
+// NewErsSorter creates a new ERSSorter
+func NewErsSorter(tablets []*topodatapb.Tablet, positions []mysql.Position, idealCell string) *ERSSorter {
+	return &ERSSorter{
+		tablets:   tablets,
+		positions: positions,
+		idealCell: idealCell,
+	}
+}
+
+// Len implements the Interface for sorting
+func (ersSorter *ERSSorter) Len() int { return len(ersSorter.tablets) }
+
+// Swap implements the Interface for sorting
+func (ersSorter *ERSSorter) Swap(i, j int) {
+	ersSorter.tablets[i], ersSorter.tablets[j] = ersSorter.tablets[j], ersSorter.tablets[i]
+	ersSorter.positions[i], ersSorter.positions[j] = ersSorter.positions[j], ersSorter.positions[i]
+}
+
+// Less implements the Interface for sorting
+func (ersSorter *ERSSorter) Less(i, j int) bool {
+	// Returning "true" in this function means [i] is before [j] in the sorting order,
```

Comment:
This ERSSorter is used in when we are trying to find the intermediate source for replication. In this case we first and foremost want the most advanced tablet even if it cannot become a primary itself, but since it is the most advanced we must use it for replication initially. Once, we have selected the intermediate source, we have a separate function `identifyPrimaryCandidate` to find the idead primary candidate which chooses the candidate in the order of the promotion rules and finally we have a constraint check in the end which will prohibit any must_not tablet reaching that point (this will only happen if we do not have any candidate that can be promoted).

I think the confusion is arising from the name of the struct `ERSSorter`. I think `ERSIntermediateSourceSorter` would be a more descriptive name but it would be a little too long. WDYT?